### PR TITLE
feat: Introduce iceberg coordinated committer

### LINKED ISSUE #10642: Implement `CoordinatedSinkWriter` for iceberg sink.

### PR DESCRIPTION
**Pull request by liurenjie1024**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Closes #10642 #10643 

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/connector/src/sink/iceberg.rs [hunk 1]
```diff
@@ -201,8 +204,7 @@ impl IcebergSink {
 
 #[async_trait::async_trait]
 impl Sink for IcebergSink {
-    /// TODO: Can't support commit coordinator now.
-    type Coordinator = DummySinkCommitCoordinator;
+    type Coordinator = IcebergSinkCommitter;
     type Writer = IcebergWriter;
```
Comment (ZENOTME) [inline src/connector/src/sink/iceberg.rs:208]: Should we wrap it with CoordinatedSinkWriter if we want to use a seperate Coordinator?

Comment (liurenjie1024) [reply src/connector/src/sink/iceberg.rs:208]: I think it's already wrapped in https://github.com/risingwavelabs/risingwave/blob/2342e8b8fc036c004e6935628e790ef007a6d6c7/src/connector/src/sink/coordinate.rs#L24

File: src/connector/Cargo.toml [hunk 2]
```diff
@@ -45,9 +45,9 @@ duration-str = "0.5.1"
 enum-as-inner = "0.6"
 futures = { version = "0.3", default-features = false, features = ["alloc"] }
 futures-async-stream = { workspace = true }
+icelake = {workspace = true}
```
Comment (wenym1) [inline src/connector/Cargo.toml:48]: nit: need format.

File: src/connector/src/sink/mod.rs [hunk 3]
```diff
@@ -130,7 +130,7 @@ pub struct SinkWriterParam {
 
 #[async_trait]
 pub trait Sink {
-    type Writer: SinkWriter<CommitMetadata = ()>;
+    type Writer: SinkWriter;
```
Comment (wenym1) [inline src/connector/src/sink/mod.rs:133]: Why the associated type bound is removed?

File: src/connector/src/sink/iceberg.rs [hunk 4]
```diff
@@ -201,8 +207,7 @@ impl IcebergSink {
...
 
     async fn validate(&self, _client: Option<ConnectorClient>) -> Result<()> {
```
Comment (wenym1) [inline src/connector/src/sink/iceberg.rs:213]: This method is for validating the user input props and will be called before the sink is created. I think we should check whether table exists and table schema and do other validation here.

File: src/connector/src/sink/iceberg.rs (hunk 1 above)
Comment (wenym1) [reply src/connector/src/sink/iceberg.rs:211]: It's not wrapped yet. It should be `type Writer = CoordinatedSinkWriter<IcebergWriter>`. The `CoordinatedSinkWriter` is responsible for communication with the coordinator.

File: src/connector/src/sink/mod.rs (hunk 3 above)
Comment (wenym1) [reply src/connector/src/sink/mod.rs:133]: If we wrap the `IcebergWriter` with the `CoordinatedSinkWriter`, we don't have to remove the associated type bound here.

File: src/connector/src/sink/iceberg.rs [hunk 5]
```diff
@@ -320,3 +379,37 @@ impl SinkWriter for IcebergWriter {
...
+
+    async fn commit(&mut self, epoch: u64, metadata: Vec<SinkMetadata>) -> Result<()> {
+        tracing::info!("Starting iceberg commit in epoch {epoch}.");
+
+        let write_results = metadata
+            .iter()
+            .map(WriteResult::try_from)
+            .collect::<Result<Vec<WriteResult>>>()?;
+
+        let mut tx = Transaction::new(&mut self.table);
```
Comment (wenym1) [inline src/connector/src/sink/iceberg.rs:402]: `txn` to avoid ambiguity with channel sender?

File: src/connector/src/sink/iceberg.rs (hunk 1 above)
Comment (liurenjie1024) [reply src/connector/src/sink/iceberg.rs]: Fixed. But this design doesn't follow DRY principal, but we can improve it later.

**Review by wenym1 — APPROVED** [review]
LGTM!

**Pull request merged**